Support for AArch64/CentOS

Try to compile the source on an aarch64 architecture with CentOS (RPM Package Manager) installed environment, and find the following error:

```
/home/malt-1.2.0/src/lib/core/CallStackInfo.hpp:62:40: error: ‘ticks’ has not been declared
   void onFreeLinkedMemory(size_t value,ticks lifetime,size_t peakId);
                                         ^~~~~
/home/malt-1.2.0/src/lib/core/CallStackInfo.cpp:166:54: error: ‘ticks’ has not been declared
 void CallStackInfo::onFreeLinkedMemory(size_t value, ticks lifetime,size_t peakId)
                                                      ^~~~~
```

I update the file and redo the cmake steps, it seems the errors are still there.

```
Scanning dependencies of target malt-core
[  2%] Building CXX object src/lib/core/CMakeFiles/malt-core.dir/VmaTracker.cpp.o
[  2%] Building CXX object src/lib/core/CMakeFiles/malt-core.dir/SymbolSolver.cpp.o
[  3%] Building CXX object src/lib/core/CMakeFiles/malt-core.dir/CallStackInfo.cpp.o
In file included from /home/b84170347/malt-master/src/lib/core/CallStackInfo.cpp:15:0:
/home/userid/malt-master/src/lib/core/CallStackInfo.hpp:62:40: error: ‘ticks’ has not been declared
   void onFreeLinkedMemory(size_t value,ticks lifetime,size_t peakId);
                                        ^~~~~
/home/userid/malt-master/src/lib/core/CallStackInfo.cpp:166:54: error: ‘ticks’ has not been declared
 void CallStackInfo::onFreeLinkedMemory(size_t value, ticks lifetime,size_t peakId)
                                                      ^~~~~
make[2]: *** [src/lib/core/CMakeFiles/malt-core.dir/build.make:89: src/lib/core/CMakeFiles/malt-core.dir/CallStackInfo.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:305: src/lib/core/CMakeFiles/malt-core.dir/all] Error 2

```

According to /proc/cpuinfo, it should be ARM V8.( `CPU architecture: 8`)
Since it keeps complaining about `ticks` has not been declared,  I just remove those `#if defined(__aarch64__) && defined(HAVE_ARMV8_PMCCNTR_EL0)` header to force it fall into one of the cases and receive `redefinition of ticks getticks()` err


```
In file included from /home/userid/malt-master/src/lib/core/CallStackInfo.hpp:17:0,
                 from /home/userid/malt-master/src/lib/core/SimpleCallStackNode.hpp:13,
                 from /home/userid/malt-master/src/lib/core/SegmentTracker.hpp:19,
                 from /home/userid/malt-master/src/lib/core/SegmentTracker.cpp:13:
/home/userid/malt-master/src/lib/../../extern-deps/from-fftw/cycle.h: In function ‘ticks getticks()’:
/home/userid/malt-master/src/lib/../../extern-deps/from-fftw/cycle.h:556:21: error: redefinition of ‘ticks getticks()’
 static inline ticks getticks(void)
                     ^~~~~~~~
In file included from /home/userid/malt-master/src/lib/core/SegmentTracker.hpp:16:0,
                 from /home/userid/malt-master/src/lib/core/SegmentTracker.cpp:13:
/home/userid/malt-master/src/lib/../../extern-deps/from-fftw/cycle.h:556:21: note: ‘ticks getticks()’ previously defined here
 static inline ticks getticks(void)
                     ^~~~~~~~
In file included from /home/userid/malt-master/src/lib/core/CallStackInfo.hpp:17:0,
                 from /home/userid/malt-master/src/lib/core/SimpleCallStackNode.hpp:13,
                 from /home/userid/malt-master/src/lib/core/SegmentTracker.hpp:19,
                 from /home/userid/malt-master/src/lib/core/SegmentTracker.cpp:13:
/home/userid/malt-master/src/lib/../../extern-deps/from-fftw/cycle.h: In function ‘double elapsed(ticks, ticks)’:
/home/userid/malt-master/src/lib/../../extern-deps/from-fftw/cycle.h:92:47: error: redefinition of ‘double elapsed(ticks, ticks)’
 #define INLINE_ELAPSED(INL) static INL double elapsed(ticks t1, ticks t0) \
                                               ^
/home/b84170347/malt-master/src/lib/../../extern-deps/from-fftw/cycle.h:562:1: note: in expansion of macro ‘INLINE_ELAPSED’
 INLINE_ELAPSED(inline)
 ^~~~~~~~~~~~~~
In file included from /home/b84170347/malt-master/src/lib/core/SegmentTracker.hpp:16:0,
                 from /home/b84170347/malt-master/src/lib/core/SegmentTracker.cpp:13:
/home/userid/malt-master/src/lib/../../extern-deps/from-fftw/cycle.h:92:47: note: ‘double elapsed(ticks, ticks)’ previously defined here
 #define INLINE_ELAPSED(INL) static INL double elapsed(ticks t1, ticks t0) \
                                               ^
/home/userid/malt-master/src/lib/../../extern-deps/from-fftw/cycle.h:562:1: note: in expansion of macro ‘INLINE_ELAPSED’
 INLINE_ELAPSED(inline)
 ^~~~~~~~~~~~~~
make[2]: *** [src/lib/core/CMakeFiles/malt-core.dir/build.make:115: src/lib/core/CMakeFiles/malt-core.dir/SegmentTracker.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:305: src/lib/core/CMakeFiles/malt-core.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
```
I have check the gcc/g++ compiler on my machine, they did not define any MACRO such as `HAVE_ARMV8_CNTVCT_EL0` or `HAVE_ARMV8_PMCCNTR_EL0`.

The automatic detection does not work well. But with mannually editing the files, I made it compile on my machine.

I checked and comfired, the second last case works on my machine and force the complier fall into that category with removing those checking MACRO. Also I have to do `#include<stdint.h>` to let the gcc compiler recognize `uint64_t ` type.


```
#if defined(__aarch64__) && !defined(HAVE_TICK_COUNTER)
#include <stdint.h>
typedef uint64_t ticks;
static inline ticks getticks(void)
{
  uint64_t Rt;
  asm volatile("mrs %0,  CNTVCT_EL0" : "=r" (Rt));
  return Rt;
}
INLINE_ELAPSED(inline)
#define HAVE_TICK_COUNTER
#endif
```

Comment 1: Is there a btter place to `include <stdint.h>` or using `long long` to replace uint64_t.

For Automatic detection parts. 
In `src/lib/CMakeLists.txt`, I revise the syntax of check compilation by using CheckCSourceRuns
FYI, the last 4 cases are ARM related, I include them here
```
# check arm execution
include(CheckCSourceRuns)
CHECK_C_SOURCE_RUNS("int main() {long Rt, Rt2 = 0; asm volatile(\"mrrc p15, 1, %0, %1, c14\" : \"=r\"(Rt), \"=r\"(Rt2)); return 0;}" HAVE_ARMV7A_CNTVCT)
CHECK_C_SOURCE_RUNS("int main() {lont r; asm volatile(\"mrc p15, 0, %0, c9, c13, 0\" : \"=r\"(r) ); return 0;}" HAVE_ARMV7A_PMCCNTR)
CHECK_C_SOURCE_RUNS("int main() {long long Rt; asm volatile(\"mrs %0,  CNTVCT_EL0\" : \"=r\" (Rt)); return 0;}" HAVE_ARMV8_CNTVCT_EL0)
CHECK_C_SOURCE_RUNS("int main() {long long cc; asm volatile(\"mrs %0, PMCCNTR_EL0\" : \"=r\"(cc)); return 0;}" HAVE_ARMV8_PMCCNTR_EL0)
```
Comment 2: `asm volatile(\"mrs %0, PMCCNTR_EL0\" : \"=r\"(cc)); return 0;}` succeeds in compilation but fails in execution. So I decide to use CheckCSourceRuns.

And add one more `cmakedefine` in the corresponding part in `src/lib/config.h.in`

```
//arm RDTSC port selection
#cmakedefine HAVE_ARMV7A_CNTVCT
#cmakedefine HAVE_ARMV7A_PMCCNTR
#cmakedefine HAVE_ARMV8_CNTVCT_EL0
#cmakedefine HAVE_ARMV8_PMCCNTR_EL0
```

Comment 3: With the chages above, I have seen there are 3 `failed` 1 `success` (Excepted Behavior)
```
-- Performing Test HAVE_ARMV7A_CNTVCT
-- Performing Test HAVE_ARMV7A_CNTVCT - Failed
-- Performing Test HAVE_ARMV7A_PMCCNTR
-- Performing Test HAVE_ARMV7A_PMCCNTR - Failed
-- Performing Test HAVE_ARMV8_CNTVCT_EL0
-- Performing Test HAVE_ARMV8_CNTVCT_EL0 - Success
-- Performing Test HAVE_ARMV8_PMCCNTR_EL0
-- Performing Test HAVE_ARMV8_PMCCNTR_EL0 - Failed

```
But the build process is still broken, with 
```
/home/malt-fix-57-aarch64-asm-detection/src/lib/core/CallStackInfo.hpp:62:40: error: ‘ticks’ has not been declared
   void onFreeLinkedMemory(size_t value,ticks lifetime,size_t peakId);
                                         ^~~~~
/homemalt-fix-57-aarch64-asm-detection/src/lib/core/CallStackInfo.cpp:166:54: error: ‘ticks’ has not been declared
 void CallStackInfo::onFreeLinkedMemory(size_t value, ticks lifetime,size_t peakId)
```
It seems the `HAVE_ARMV8_CNTVCT_EL0` has not been defined (I revert my manual changes)
Although I have seen its defination in  `/home/userid/malt-fix-57-aarch64-asm-detection/build/src/lib/config.h
`
```
//arm RDTSC port selection
/* #undef HAVE_ARMV7A_CNTVCT */
/* #undef HAVE_ARMV7A_PMCCNTR */
#define HAVE_ARMV8_CNTVCT_EL0
/* #undef HAVE_ARMV8_PMCCNTR_EL0 */

```

I guess that is related to the order of including? Can you check this issues?

--- a/ChangeLog.md
+++ b/ChangeLog.md
@@ -3,6 +3,7 @@ Change log
 
 master
 ------
+ - Build support for aarch64
 
 1.2.0 - 28/03/2019
 ------------------

--- a/cmake/FindLibunwind.cmake
+++ b/cmake/FindLibunwind.cmake
@@ -18,7 +18,7 @@ find_library(LIBUNWIND_LIBRARY NAMES unwind
 	HINTS ${LIBUNWIND_PREFIX}/lib)
 
 ######################################################
-foreach(arch x86 x86_64 ppc ppc64)
+foreach(arch x86 x86_64 ppc ppc64 aarch aarch64)
 	find_library(LIBUNWIND_${arch} NAMES unwind-${arch}
 		HINTS ${LIBUNWIND_PREFIX}/lib)
 	if (LIBUNWIND_${arch})

--- a/extern-deps/from-fftw/cycle.h
+++ b/extern-deps/from-fftw/cycle.h
@@ -78,6 +78,9 @@
 
 /***************************************************************************/
 
+//config
+#include <config.h>
+
 #if TIME_WITH_SYS_TIME
 # include <sys/time.h>
 # include <time.h>
@@ -515,7 +518,8 @@ INLINE_ELAPSED(inline)
 #endif
 #endif /* HAVE_MIPS_ZBUS_TIMER */
 
-#if defined(HAVE_ARMV7A_CNTVCT)
+#if defined(HAVE_ARMV7A_CNTVCT) && !defined(HAVE_TICK_COUNTER)
+#include <stdint.h>
 typedef uint64_t ticks;
 static inline ticks getticks(void)
 {
@@ -528,6 +532,7 @@ INLINE_ELAPSED(inline)
 #endif
 
 #if defined(HAVE_ARMV7A_PMCCNTR) && !defined(HAVE_TICK_COUNTER)
+#include <stdint.h>
 typedef uint64_t ticks;
 static inline ticks getticks(void)
 {
@@ -540,6 +545,7 @@ INLINE_ELAPSED(inline)
 #endif
 
 #if defined(__aarch64__) && defined(HAVE_ARMV8_CNTVCT_EL0) && !defined(HAVE_ARMV8_PMCCNTR_EL0) && !defined(HAVE_TICK_COUNTER)
+#include <stdint.h>
 typedef uint64_t ticks;
 static inline ticks getticks(void)
 {
@@ -552,6 +558,7 @@ INLINE_ELAPSED(inline)
 #endif
 
 #if defined(__aarch64__) && defined(HAVE_ARMV8_PMCCNTR_EL0) && !defined(HAVE_TICK_COUNTER)
+#include <stdint.h>
 typedef uint64_t ticks;
 static inline ticks getticks(void)
 {

--- a/src/lib/CMakeLists.txt
+++ b/src/lib/CMakeLists.txt
@@ -55,6 +55,14 @@ if (LIBUNWIND_FOUND)
 endif (LIBUNWIND_FOUND)
 
 ######################################################
+# check arm compilation
+include(CheckCXXSourceCompiles)
+CHECK_CXX_SOURCE_COMPILES("uint32_t r; asm volatile(\"mrc p15, 0, %0, c9, c13, 0\" : \"=r\"(r) );" HAVE_ARMV7A_PMCCNTR)
+CHECK_CXX_SOURCE_COMPILES("int64_t Rt; asm volatile(\"mrs %0,  CNTVCT_EL0\" : \"=r\" (Rt));" HAVE_ARMV8_CNTVCT_EL0)
+CHECK_CXX_SOURCE_COMPILES("uint64_t cc = 0; asm volatile(\"mrs %0, PMCCNTR_EL0\" : \"=r\"(cc));" HAVE_ARMV8_PMCCNTR_EL0)
+
+
+######################################################
 #gen config.h
 configure_file(${CMAKE_CURRENT_SOURCE_DIR}/config.h.in ${CMAKE_CURRENT_BINARY_DIR}/config.h)
 

--- a/src/lib/CMakeLists.txt
+++ b/src/lib/CMakeLists.txt
@@ -56,10 +56,12 @@ endif (LIBUNWIND_FOUND)
 
 ######################################################
 # check arm compilation
-include(CheckCXXSourceCompiles)
-CHECK_CXX_SOURCE_COMPILES("uint32_t r; asm volatile(\"mrc p15, 0, %0, c9, c13, 0\" : \"=r\"(r) );" HAVE_ARMV7A_PMCCNTR)
-CHECK_CXX_SOURCE_COMPILES("int64_t Rt; asm volatile(\"mrs %0,  CNTVCT_EL0\" : \"=r\" (Rt));" HAVE_ARMV8_CNTVCT_EL0)
-CHECK_CXX_SOURCE_COMPILES("uint64_t cc = 0; asm volatile(\"mrs %0, PMCCNTR_EL0\" : \"=r\"(cc));" HAVE_ARMV8_PMCCNTR_EL0)
+# check arm execution
+include(CheckCSourceRuns)
+CHECK_C_SOURCE_RUNS("int main() {long Rt, Rt2 = 0; asm volatile(\"mrrc p15, 1, %0, %1, c14\" : \"=r\"(Rt), \"=r\"(Rt2)); return 0;}" HAVE_ARMV7A_CNTVCT)
+CHECK_C_SOURCE_RUNS("int main() {lont r; asm volatile(\"mrc p15, 0, %0, c9, c13, 0\" : \"=r\"(r) ); return 0;}" HAVE_ARMV7A_PMCCNTR)
+CHECK_C_SOURCE_RUNS("int main() {long long Rt; asm volatile(\"mrs %0,  CNTVCT_EL0\" : \"=r\" (Rt)); return 0;}" HAVE_ARMV8_CNTVCT_EL0)
+CHECK_C_SOURCE_RUNS("int main() {long long cc; asm volatile(\"mrs %0, PMCCNTR_EL0\" : \"=r\"(cc)); return 0;}" HAVE_ARMV8_PMCCNTR_EL0)
 
 
 ######################################################

--- a/src/lib/config.h.in
+++ b/src/lib/config.h.in
@@ -30,4 +30,9 @@
 //if have libunwind
 #cmakedefine MALT_HAVE_LIBUNWIND
 
+//arm RDTSC port selection
+#cmakedefine HAVE_ARMV7A_PMCCNTR
+#cmakedefine HAVE_ARMV8_CNTVCT_EL0
+#cmakedefine HAVE_ARMV8_PMCCNTR_EL0
+
 #endif //CONFIG_H

--- a/src/lib/config.h.in
+++ b/src/lib/config.h.in
@@ -31,6 +31,7 @@
 #cmakedefine MALT_HAVE_LIBUNWIND
 
 //arm RDTSC port selection
+#cmakedefine HAVE_ARMV7A_CNTVCT
 #cmakedefine HAVE_ARMV7A_PMCCNTR
 #cmakedefine HAVE_ARMV8_CNTVCT_EL0
 #cmakedefine HAVE_ARMV8_PMCCNTR_EL0

--- a/src/lib/core/StackSizeTracker.cpp
+++ b/src/lib/core/StackSizeTracker.cpp
@@ -9,6 +9,7 @@
 /********************  HEADERS  *********************/
 #include <cassert>
 #include <cstdio>
+#include <cstdint>
 #include "StackSizeTracker.hpp"
 #include <portability/LinuxProcMapReader.hpp>
 #include <cycle.h>
@@ -32,16 +33,15 @@ StackSizeTracker::StackSizeTracker(void)
 void StackSizeTracker::enter(void)
 {
 	//get stack size
-	unsigned long crbp;
 	unsigned long crsp;
 	
 	//read counters
-	#ifdef __x86_64__
-		asm("movq %%rbp,%0" : "=r"(crbp));
+	#if defined(__x86_64__)
 		asm("movq %%rsp,%0" : "=r"(crsp));
-	#elif __i386__
-		asm("mov %%ebp,%0" : "=r"(crbp));
+	#elif defined(__i386__)
 		asm("mov %%esp,%0" : "=r"(crsp));
+	#elif defined(__aarch__) || defined(__aarch64__)
+		asm("mov %0, sp" : "=r"(crsp));
 	#else
 		MALT_FATAL("Enter-exit mode is not supported for this architecture !");
 		#warning "Arch not supported, stack size tracking will be ignored, all rest will work fine, this is optional."

--- a/src/lib/tests/CMakeLists.txt
+++ b/src/lib/tests/CMakeLists.txt
@@ -44,11 +44,9 @@ add_executable(simple-case-no-finstr simple-case.cpp)
 
 ######################################################
 add_executable(simple-case-finstr simple-case.cpp)
-SET_TARGET_PROPERTIES(simple-case-finstr PROPERTIES COMPILE_FLAGS -finstrument-functions)
 
 ######################################################
 add_executable(simple-case-finstr-linked simple-case.cpp)
-SET_TARGET_PROPERTIES(simple-case-finstr-linked PROPERTIES COMPILE_FLAGS -finstrument-functions)
 target_link_libraries(simple-case-finstr-linked malt)
 
 ######################################################

--- a/src/lib/tools/ELFReader.cpp
+++ b/src/lib/tools/ELFReader.cpp
@@ -21,7 +21,7 @@ namespace MALT
 {
 
 /********************  MACROS  **********************/
-#ifdef __x86_64__
+#if defined(__x86_64__) || defined(__aarch64__)
 	#define ElfArch_Ehdr Elf64_Ehdr
 	#define elfarch_getehdr(x) elf64_getehdr(x)
 	#define ElfArch_Shdr Elf64_Shdr
@@ -29,7 +29,7 @@ namespace MALT
 	#define ElfArch_Sym Elf64_Sym
 	#define ELFARCH_ST_BIND(x) ELF64_ST_BIND(x)
 	#define ELFARCH_ST_TYPE(x) ELF64_ST_TYPE(x)
-#elif __i386__
+#elif defined(__i386__) || defined(__aarch__)
 	#define ElfArch_Ehdr Elf32_Ehdr
 	#define elfarch_getehdr(x) elf32_getehdr(x)
 	#define ElfArch_Shdr Elf32_Shdr

--- a/src/trace-reader/CMakeLists.txt
+++ b/src/trace-reader/CMakeLists.txt
@@ -8,6 +8,7 @@
 
 ######################################################
 include_directories(../lib)
+include_directories(${CMAKE_CURRENT_BINARY_DIR}/../lib)
 include_directories(../../extern-deps/from-fftw)
 
 ######################################################

